### PR TITLE
Fix the compiler so it handles ambiguous `foo[bar]` type syntax

### LINKED ISSUE #1054: bool array with enum_count
It seems like `enum_count` still doesn't work in array sizes. Consider this code, taken from curses-klondike:

```python
class ParseState:
    program_name: byte*
    remaining_args: byte**  # never includes program name
    # TODO(Jou): This doesn't work yet. For now let's add much more than enough...
    #seen: bool[enum_count(Option)]
    seen: bool[10]
    result: Args
```

If I delete `seen: bool[10]` and uncomment the commented out code, it fails:

```
compiler error in file "src/args.jou", line 45: expected a type, got the 'enum_count' keyword
```

#1051 does not help with this.

### PR DESCRIPTION
**Pull request by Akuli**
The type `foo[bar]` can be either an array of `foo` whose length is `bar`, or a generic class `foo` with parameter `bar`. This pull request properly implements everything that the compiler needs to know about this ambiguity. If I did this right, it can now figure out the ambiguity in all possible cases.

Checks a couple boxes in #1045. Fixes #1054. Affects #845 but doesn't fix it.

### CODE REVIEW

File: compiler/tokenizer.jou [hunk 1]
```diff
@@ -15,6 +15,7 @@ def is_identifier_or_number_byte(b: byte) -> bool:
 def is_operator_byte(c: byte) -> bool:
     return c != '\0' and strchr("=<>!.,()[]{};:+-*/&%|^~", c) != NULL
 
+@public
```
Comment (Akuli) [inline compiler/tokenizer.jou]: nah
```suggestion
```

File: compiler/typecheck/common.jou [hunk 2]
```diff
@@ -23,6 +25,75 @@ def plural_s(n: int64) -> byte*:
...
+            }
+        case AstTypeKind.Array:
+            # TODO: implement this
+            fail(type.location, "cannot evaluate array length at compile time")
+        case AstTypeKind.Generic:
+            # TODO: implement this
+            fail(type.location, "cannot evaluate array length at compile time")
+        case AstTypeKind.Pointer | AstTypeKind.FuncPtr | AstTypeKind.BuiltIn:
+            # "Foo*" or "funcptr(...) -> ..." or "int" is not a valid expression
+            # TODO: test this
```
Comment (Akuli) [inline compiler/typecheck/common.jou]: Tested now
```suggestion
```

File: compiler/typecheck/common.jou [hunk 3]
```diff
@@ -23,6 +25,74 @@ def plural_s(n: int64) -> byte*:
         return "s"
 
 
+def array_type_from_ast(jou_file: JouFile*, item_type: Type*, length_ast: AstExpression*) -> Type*:
```
Comment (Akuli) [inline compiler/typecheck/common.jou]: Seems to be called in only one place, inline?

File: compiler/typecheck/common.jou [hunk 4]
```diff
@@ -23,6 +25,74 @@ def plural_s(n: int64) -> byte*:
...
+# The syntax is ambiguous, and the parser assumes generic rather than array.
+# This function tries to interpret the type Foo[Bar] as an array.
+def handle_bad_generic(jou_file: JouFile*, generic: AstGenericType*, location: Location) -> Type*:
+    assert generic != NULL
+    name: byte* = generic.class_name
+    msg: byte[500]
+
+    item_type = find_and_typecheck_type(jou_file, name)
+    if item_type == NULL:
+        # TODO: make sure both error messages are tested
```
Comment (Akuli) [inline compiler/typecheck/common.jou]: Must do

File: tests/syntax_error/array_length_from_type.jou [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# This is a funny corner case. The compiler first interprets Foo[x*] as a
+# generic class, but then realizes that Foo is not generic and attempts to
+# interpret x* as an expression instead of a type.
+
+class Foo:
+    pass
+
+def blah() -> None:
+    x = 123
+    y: Foo[x*]  # Error: invalid syntax in array length
```
Comment (Akuli) [inline tests/syntax_error/array_length_from_type.jou]: Not the best error message, maybe:


```suggestion
    y: Foo[x*]  # Error: invalid array length
```
